p2p/src/sync/mod.rs

```diff
@@ -173,40 +157,59 @@ where
-    ///
-    /// The `HeaderListRequest` message is sent to newly connected peers.
-    async fn register_peer(&mut self, peer: PeerId) -> Result<()> {
+    /// Starts a task for the new peer.
+    pub fn register_peer(&mut self, peer: PeerId) -> Result<()> {
         log::debug!("Register peer {peer} to sync manager");
 
-        self.request_headers(peer).await?;
-        match self.peers.insert(peer, PeerContext::new()) {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        self.peers
+            .insert(peer, sender)
             // This should never happen because a peer can only connect once.
-            Some(_) => Err(P2pError::PeerError(PeerError::PeerAlreadyExists)),
-            None => Ok(()),
-        }
+            .map(|_| Err::<(), _>(P2pError::PeerError(PeerError::PeerAlreadyExists)))
+            .transpose()?;
+
+        let peer_sender = self.peer_sender.clone();
+        let peer_manager_sender = self.peer_manager_sender.clone();
+        let chainstate_handle = self.chainstate_handle.clone();
+        let p2p_config = Arc::clone(&self.p2p_config);
+        let is_initial_block_download = Arc::clone(&self.is_initial_block_download);
+        tokio::spawn(async move {
+            let mut peer = Peer::<T>::new(
+                peer,
+                p2p_config,
+                chainstate_handle,
+                peer_manager_sender,
+                peer_sender,
+                receiver,
+                is_initial_block_download,
+            );
+            if let Err(e) = peer.run().await {
+                log::error!("Sync manager peer ({}) error: {e:?}", peer.id());
+            }
+        });
+
+        Ok(())
     }
 
-    /// Removes the state (`PeerContext`) of the given peer.
+    /// Stops the task of the given peer by closing the corresponding channel.
     fn unregister_peer(&mut self, peer: PeerId) {
         log::debug!("Unregister peer {peer} from sync manager");
 
-        // Remove the queued block responses associated with the disconnected peer.
-        self.blocks_queue.retain(|(p, _)| p != &peer);
-
-        if self.peers.remove(&peer).is_some() {
+        if self.peers.remove(&peer).is_none() {
             log::warn!("Unregistering unknown peer: {peer}");
         }
     }
 
     /// Announces the header of a new block to peers.
     async fn handle_new_tip(&mut self, block_id: Id<Block>) -> Result<()> {
-        if self.is_initial_block_download {
-            self.is_initial_block_download =
-                self.chainstate_handle.call(|c| c.is_initial_block_download()).await??;
+        if self.is_initial_block_download.load(Ordering::SeqCst) {
+            self.is_initial_block_download.store(
+                self.chainstate_handle.call(|c| c.is_initial_block_download()).await??,
+                Ordering::Release,
+            );
         }
 
-        if self.is_initial_block_download {
+        if self.is_initial_block_download.load(Ordering::SeqCst) {
```

Comment:
Also `Ordering::Acquire`?

Reply:
I would even use `Ordering::Relaxed` for this, because at this point, the worst is missing one block from this condition. The best is removing all synchronization primitives that will be done on every new tip.

Reply:
I think calling this twice is a waste... and is bad practice. You could do this:

```
        let is_initial_block_download = if self.is_initial_block_download.load(Ordering::Relaxed) {
            let chainstate_ibd =
                self.chainstate_handle.call(|c| c.is_initial_block_download()).await??;
            self.is_initial_block_download.store(chainstate_ibd, Ordering::Release);
            chainstate_ibd
        } else {
            false
        };

        if is_initial_block_download {
            return Ok(());
        }
```